rust/gitxetcore/src/git_integration/repo_migration.rs

```diff
@@ -0,0 +1,675 @@
+
+        for t_oid in trees_to_convert.iter() {
+            let tree = src.find_tree(*t_oid)?;
+
+            for item in tree.iter() {
+                if matches!(item.kind(), Some(ObjectType::Blob)) {
+                    filterable_blobs.insert(item.id());
+                }
+            }
+        }
+
+        // Now, run the bulk of the blob processing in parallel as
+        let blob_processing_permits = Arc::new(Semaphore::new(MAX_CONCURRENT_BLOB_PROCESSING));
+        let mut blob_processing_pool = JoinSet::<Result<(Oid, Vec<u8>)>>::new();
+
+        // Now, with everything there, first go through and convert all the blobs.
+        for b_oid in blobs_to_convert {
+            if filterable_blobs.contains(&b_oid) {
+                // Add this to the filtering pool.
+                let blob_processing_permits = blob_processing_permits.clone();
+                let progress_reporting = progress_reporting.clone();
+                let src = src.clone();
+                let pft = pft.clone();
+                let src_repo = src_repo.clone();
+                let repo_init_tracking = repo_init_tracking.clone();
+
+                blob_processing_pool.spawn(async move {
+                    let _permit = blob_processing_permits.acquire_owned().await?;
+
+                    let src_data = src.find_blob(b_oid)?.content().to_vec();
+
+                    let new_data = translate_blob_contents(
+                        &src_repo,
+                        pft.clone(),
+                        b_oid,
+                        progress_reporting,
+                        src_data,
+                        repo_init_tracking,
+                    )
+                    .await?;
+                    Ok((b_oid, new_data))
+                });
+            } else {
+                let src_blob = src.find_blob(b_oid)?;
+                let content = src_blob.content();
+                let new_oid = dest.blob(content)?;
+                tr_map.insert(b_oid, new_oid);
+                progress_reporting.register_progress(Some(1), Some(content.len()));
+            }
+
+            while let Some(res) = blob_processing_pool.try_join_next() {
+                let (b_oid, new_data) = res??;
+                let new_id = dest.blob(&new_data[..])?;
+
+                tr_map.insert(b_oid, new_id);
+                progress_reporting.register_progress(Some(1), None);
+            }
+        }
+
+        // Now, clear out the rest.
+        while let Some(res) = blob_processing_pool.join_next().await {
+            let (b_oid, new_data) = res??;
+            let new_id = dest.blob(&new_data[..])?;
+
+            tr_map.insert(b_oid, new_id);
+            progress_reporting.register_progress(Some(1), None);
+        }
+
+        pft.finalize().await?;
+        progress_reporting.finalize();
+    }
+
+    // Create a gitattributes blob object, if it's not there already.
+    let gitattributes_oid = dest.blob(GITATTRIBUTES_CONTENT.as_bytes())?;
```

Comment:
Does this function return error if it's already there?

Reply:
No, if it's already present there isn't an issue.